pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTest.java

```diff
@@ -0,0 +1,263 @@
+            futureList.add(mlPendingAckStoreForWrite.appendCommitMark(txnID, CommandAck.AckType.Cumulative));
+        }
+        for (int i = 10; i < 20; i++){
+            TxnID txnID = new TxnID(i, i);
+            futureList.add(mlPendingAckStoreForWrite.appendAbortMark(txnID, CommandAck.AckType.Cumulative));
+        }
+        for (int i = 40; i < 50; i++){
+            TxnID txnID = new TxnID(i, i);
+            PositionImpl position = PositionImpl.get(i, i);
+            futureList.add(mlPendingAckStoreForWrite.appendCumulativeAck(txnID, position));
+        }
+        FutureUtil.waitForAll(futureList).get();
+        // Verify build sparse indexes correct after add many cmd-ack.
+        ArrayList<Long> positionList = new ArrayList<>();
+        for (long i = 0; i < 50; i++){
+            positionList.add(i);
+        }
+        // The indexes not contains the data which is commit or abort.
+        LinkedHashSet<Long> skipSet = new LinkedHashSet<>();
+        for (long i = 20; i < 40; i++){
+            skipSet.add(i);
+        }
+        if (writeWithBatch) {
+            for (long i = 0; i < 50; i++){
+                if (i % 2 == 0){
+                    // The indexes contains only the last position in the batch.
+                    skipSet.add(i);
+                }
+            }
+        }
+        LinkedHashSet<Long> expectedPositions = calculatePendingAckIndexes(positionList, skipSet);
+        Assert.assertEquals(
+                mlPendingAckStoreForWrite.pendingAckLogIndex.keySet().stream()
+                        .map(PositionImpl::getEntryId).collect(Collectors.toList()),
+                new ArrayList<>(expectedPositions)
+        );
+        // Replay.
+        TxnLogBufferedWriterConfig configForReplay = new TxnLogBufferedWriterConfig();
+        configForReplay.setBatchEnabled(readWithBatch);
+        configForReplay.setBatchedWriteMaxRecords(2);
+        // Denied scheduled flush.
+        configForReplay.setBatchedWriteMaxDelayInMillis(1000 * 3600);
+        MLPendingAckStore mlPendingAckStoreForRead = createPendingAckStore(configForReplay);
+        PendingAckHandleImpl pendingAckHandle = mock(PendingAckHandleImpl.class);
+        when(pendingAckHandle.getInternalPinnedExecutor()).thenReturn(internalPinnedExecutor);
+        when(pendingAckHandle.changeToReadyState()).thenReturn(true);
+        // Process controller, mark the replay task already finish.
+        final AtomicInteger processController = new AtomicInteger();
+        doAnswer(new Answer() {
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+                processController.incrementAndGet();
+                return null;
+            }
+        }).when(pendingAckHandle).completeHandleFuture();
+        mlPendingAckStoreForRead.replayAsync(pendingAckHandle, internalPinnedExecutor);
+        Awaitility.await().atMost(200, TimeUnit.SECONDS).until(() -> processController.get() == 1);
```

Comment:
Why need to wait at most 200 seconds? we don't have too many logs in the pending ack store.

Reply:
Already change to 2 seconds.